File: integrations/android/app/src/main/java/app/airplaystatus/alwayson/AlwaysOnState.kt
package app.airplaystatus.alwayson

/**
 * Pure always-on display state machine for the P7 Android client.
 *
 * Mirrors the shared rules in
 * specs/guidelines/always-on-display-client.md (and the JS/DeskThing version in
 * integrations/deskthing/shared/alwaysOnState.js). Kept free of Android APIs so
 * it runs as plain JVM unit tests (see AlwaysOnStateTest).
 */

enum class Mode { PLAYING, PAUSED, IDLE }
enum class Screen { AWAKE, DIM }

data class AlwaysOnState(
    val mode: Mode = Mode.IDLE,
    val screen: Screen = Screen.AWAKE,
    val focusBeforeIdle: Boolean = false,
    val nudge: Boolean = false,
)

sealed interface AlwaysOnEvent {
    /** A fresh /api/status snapshot. */
    data class Playback(val isPlaying: Boolean, val hasTitle: Boolean) : AlwaysOnEvent
    /** The idle grace period elapsed; `focused` = was this app foreground now. */
    data class IdleGraceElapsed(val focused: Boolean) : AlwaysOnEvent
    /** The user left the app before it dimmed. */
    data object FocusLost : AlwaysOnEvent
    /** The user tapped the resume nudge. */
    data object DismissNudge : AlwaysOnEvent
}

fun classify(isPlaying: Boolean, hasTitle: Boolean): Mode = when {
    !isPlaying && !hasTitle -> Mode.IDLE
    isPlaying -> Mode.PLAYING
    else -> Mode.PAUSED
}

fun reduce(state: AlwaysOnState, event: AlwaysOnEvent): AlwaysOnState = when (event) {
    is AlwaysOnEvent.Playback -> {
        val mode = classify(event.isPlaying, event.hasTitle)
        when {
            mode == Mode.IDLE -> state.copy(mode = Mode.IDLE, nudge = false)
            state.screen == Screen.DIM && state.focusBeforeIdle ->
                state.copy(mode = mode, nudge = true) // resume nudge; wait for tap
            state.screen == Screen.DIM ->
                state.copy(mode = mode, screen = Screen.AWAKE, nudge = false) // silent wake
            else -> state.copy(mode = mode, screen = Screen.AWAKE, nudge = false)
        }
    }

    is AlwaysOnEvent.IdleGraceElapsed ->
        if (state.mode != Mode.IDLE) state
        else state.copy(screen = Screen.DIM, focusBeforeIdle = event.focused)

    AlwaysOnEvent.FocusLost ->
        if (state.screen == Screen.DIM) state else state.copy(focusBeforeIdle = false)

    AlwaysOnEvent.DismissNudge ->
        state.copy(screen = Screen.AWAKE, nudge = false)
}

/** Whether the Activity should hold FLAG_KEEP_SCREEN_ON right now. */
fun shouldKeepScreenOn(state: AlwaysOnState): Boolean =
    state.screen == Screen.AWAKE && (state.mode == Mode.PLAYING || state.mode == Mode.PAUSED)

File: integrations/deskthing/shared/alwaysOnState.js
/**
 * Shared always-on display state machine.
 *
 * Implements the product rules from
 * ../../../specs/guidelines/always-on-display-client.md as a pure, framework-free
 * reducer so the same logic can be unit-tested and reused by the DeskThing app
 * (P8). The P7 Android client mirrors this logic in Kotlin; the P9 web client
 * uses the browser equivalent in src/public/js/displayState.js.
 *
 * State: { mode, screen, focusBeforeIdle, nudge }
 *   mode:            'playing' | 'paused' | 'idle'
 *   screen:          'awake' | 'dim'      (intent; platform maps to backlight/sleep)
 *   focusBeforeIdle: was this client the focused console when it dimmed?
 *   nudge:           show the "tap to resume" affordance right now?
 */

/** @param {{isPlaying?:boolean, title?:string|null}|null|undefined} playback */
export const classifyPlayback = (playback) => {
  if (!playback || (!playback.isPlaying && !playback.title)) return 'idle';
  return playback.isPlaying ? 'playing' : 'paused';
};

export const initialState = () => ({
  mode: 'idle',
  screen: 'awake',
  focusBeforeIdle: false,
  nudge: false,
});

/**
 * @param {ReturnType<typeof initialState>} state
 * @param {{type:'playback', playback?:object}
 *        | {type:'idleGraceElapsed', focused?:boolean}
 *        | {type:'focusLost'}
 *        | {type:'dismissNudge'}} event
 */
export const reduce = (state, event) => {
  switch (event.type) {
    case 'playback': {
      const mode = classifyPlayback(event.playback);
      if (mode === 'idle') {
        return { ...state, mode: 'idle', nudge: false };
      }
      // Active again. If we had dimmed off and this client was the focused
      // session, raise a one-shot resume nudge; otherwise wake silently.
      if (state.screen === 'dim') {
        return state.focusBeforeIdle
          ? { ...state, mode, nudge: true }
          : { ...state, mode, screen: 'awake', nudge: false };
      }
      return { ...state, mode, screen: 'awake', nudge: false };
    }

    case 'idleGraceElapsed': {
      if (state.mode !== 'idle') return state;
      return { ...state, screen: 'dim', focusBeforeIdle: Boolean(event.focused) };
    }

    case 'focusLost': {
      // Leaving the app before it dims clears the "focused-before-idle" intent.
      if (state.screen === 'dim') return state;
      return { ...state, focusBeforeIdle: false };
    }

    case 'dismissNudge':
      return { ...state, screen: 'awake', nudge: false };

    default:
      return state;
  }
};

/** Whether the platform should keep the screen on right now. */
export const shouldKeepScreenOn = (state) =>
  state.screen === 'awake' && (state.mode === 'playing' || state.mode === 'paused');
